channels/pjsip/dialplan_functions.c

```diff
@@ -1784,3 +1228,71 @@ int pjsip_acf_session_refresh_write(struct ast_channel *chan, const char *cmd, c
 
 	return ast_sip_push_task_wait_serializer(channel->session->serializer, refresh_write_cb, &rdata);
 }
+
+struct hangup_data {
+	struct ast_sip_session *session;
+	int response_code;
+};
+
+static int pjsip_hangup(void *obj)
+{
+	struct hangup_data *hdata = obj;
+	pjsip_tx_data *packet = NULL;
+
+	if ((hdata->session->inv_session->state != PJSIP_INV_STATE_DISCONNECTED) &&
+		(pjsip_inv_answer(hdata->session->inv_session, hdata->response_code, NULL, NULL, &packet) == PJ_SUCCESS)) {
```

Comment:
What's the difference between using pjsip_inv_answer here and pjsip_inv_end_session?

Reply:
pjsip_inv_end_session just calls pjsip_inv_answer when we're a UAS and the state is PJSIP_INV_STATE_INCOMING (which we've already tested for).